### PR TITLE
Move assign category ids to beforeToHtml instead inside load collection

### LINKED ISSUE #32268: Breaking BC changes in the product list block
### Summary (*)
The `\Magento\Catalog\Block\Product\ListProduct::getLoadedProductCollection` is breaking BC (feel free to challenge it) in the 2.4 version.
In 2.4 the method looks like this:

```
    public function getLoadedProductCollection()
    {
        $collection = $this->_getProductCollection();

        $categoryId = $this->getLayer()->getCurrentCategory()->getId();
        foreach ($collection as $product) {
            $product->setData('category_id', $categoryId);
        }

        return $collection;
    }
```

In 2.3 and before it looked like this:

```
    public function getLoadedProductCollection()
    {
        return $this->_getProductCollection();
    }
```
The collection iteration inside the method in 2.4 causes the collection to load the db data so any after plugin or around plugin that calls the original method and tries to affect the collection does nothing to it in 2.4.

### Examples (*)

Let's consider this plugin.
```
public function afterGetLoadedProductCollection(
    \Magento\Catalog\Block\Product\ListProduct $subject, 
    \Magento\Eav\Model\Entity\Collection\AbstractCollection $collection
) {
    return $collection->addAttributeToFilter('some_attribute_code', 'some_value');
}
```
In 2.3 and before, this would cause the product list to filter only products that have the value `some_value` for the attribute with code `some_attribute_code` (beside the default filters).
In 2.4 this filter has no effect because the collection is already loaded when adding the filter.
### Proposed solution
Move the assign operation of category ids to products to the `_beforeToHtml` method in the same class after 
```
        if (!$collection->isLoaded()) {
            $collection->load();
        }
```
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

The main purpose move assign category ids operation to `_beforeToHtml`. Clean up `getLoadedProductCollection
` like before version 2.3. Make sure we get  predictable raw data collection product (any plugin from core or 3rd-party supplier can use this entry point for get product collection data from block list). Then before render data html to browser we set data `category_id` to product if data available

Here is commit previous change
https://github.com/magento/magento2/commit/03a6f4d7612c65259bad94fdf2da1feec7916c1c
https://github.com/magento/magento2/commit/baebe46dee141a95bf67695ee1bbf4b30af49018
### Fixed Issues (if relevant)

1. Fixes magento/magento2#32268

### Manual testing scenarios (*)

1. Visit any category page have product lists
2. Go to details view one specific product (Should work without any problem)

### Questions or comments

CC: @tzyganu
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by mrtuvn** [comment]
@magento run all tests

File: dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php [hunk 1]
```diff
@@ -39,7 +43,7 @@ public function testGetLoadedProductCollection()
     {
         $this->_block->setShowRootCategory(true);
         $collection = $this->_block->getLoadedProductCollection();
-        $this->assertInstanceOf(\Magento\Catalog\Model\ResourceModel\Product\Collection::class, $collection);
+        $this->assertInstanceOf(\Magento\Eav\Model\Entity\Collection\AbstractCollection::class, $collection);
```
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php]: Why did you change it?

File: dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php [hunk 2]
```diff
@@ -39,7 +43,7 @@ public function testGetLoadedProductCollection()
...
         /* Check that root category was defined for Layer as current */
         $this->assertEquals(2, $this->_block->getLayer()->getCurrentCategory()->getId());
```
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php:44]: Could you add check that collection not loaded?

File: dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php [hunk 3]
```diff
@@ -39,7 +43,7 @@ public function testGetLoadedProductCollection()
...
     }
```
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php:45]: Could you add one more test case that will make sure that category_id was set before to HTML, maybe using reflection to call before to HTML method?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php (hunk 1 above)
Comment (mrtuvn) [reply dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php]: The reason i think we should check with corresponding type return 
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Block/Product/ListProduct.php#L140
This phpdoc return AbstractCollection abstract class

File: dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php (hunk 3 above)
Comment (mrtuvn) [reply dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php:45]: I'm not sure how to coverage that when this method is protected and only can callable in origin class

Comment (mrtuvn) [reply dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php:45]: Also i not sure the part previously someone try to set `category_id` to product before render to html. Seem not well-write implements

File: dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php (hunk 1 above)
Comment (ihor-sviziev) [reply dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php]: That’s a bug in the php doc block that not so easy to fix due to SVC failures. Method getProductCollection should return product collection, so previously it was correct

File: dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php (hunk 3 above)
Comment (mrtuvn) [reply dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php:45]: I'm not sure the way to check category_id after `_beforeToHtml` called. This out of scope of my knowledge on this part
Must have some reason after block list rendered and product set category_id! Maybe later magento will need category_id for some check on products

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Hi @vzabaznov @mslabko @sidolov @sivaschenko,
Could you provide some info on why the setting category_id was added in https://github.com/magento/magento2/commit/baebe46dee141a95bf67695ee1bbf4b30af49018? Can it be moved to `beforeToHtml` as suggested in this PR?
Unfortunately, there no clear description and test coverage for these changes.

**Comment by ihor-sviziev** [comment]
From @sivaschenko  in slack:
> This was a perfomance improvement for EE edition
> I don’t think the change will break anything, but I will check with the author of the fix to make sure
> The idea was to avoid calls to resource consuming $this->getCategoryIds() on each product

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved

Note for QA: Please make sure this PR isn't introducing regression for performance improvement for the EE edition that was done in the scope of MC-33086.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8913](https://jira.corp.magento.com/browse/ENGCOM-8913) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
Moving back to "accept" status. It looks like it was moved by some mistake.
